client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+
+## HTTP Message Formatting
+First, we specify the default and custom headers our HTTP messages utilize.  Then we describe the expected forms of requests/responses between the cache and TreodeDB server.
+
+### Use of HTTP Headers
+
+In our use of HTTP headers, the term *age* refers to time since the user-specified read\_timestamp.  Specifically, every TreodeDB entry has both a *value timestamp (V_t)* and *read timestamp (R_t)*.  We maintain as an invariant that the given key contained the entry's value between [V\_t, R\_t].
+
+We use the following HTTP headers:
+
+* **max-age**: The max number of seconds between our given read\_timestamp and the read\_timestamp of the entry, i.e., (input read\_timestamp) - (entry read\_timestamp) <= max-age.  In other words, we are willing to read a value that could be stale by at most max-age seconds. 
+* **no-cache**: Forces cache to send a request to the DB server to retrieve the requested value.  The result cannot just be returned from cache.
+* **Read-TxClock**: Contains the user's specified read\_timestamp.  The value returned by the DB server will be the most recent value older than the specified read\_timestamp value.  Note that Read-TxClock is an integer; it is specified in milliseconds to clarify that it should **not** be manipulated as a date.
+
+Additionally, note that max-age and no-cache are existing HTTP Caching headers specified in [RFC7234](http://tools.ietf.org/html/rfc7234#section-5.2.1), while the read\_timestamp is sent through the Read-TxClock header, a custom HTTP header.
+
+Below are the expected forms of requests/responses between the cache and TreodeDB server:
+
+### Requests
+
+We give examples of HTTP requests below.  For additional explanation, please see the source of these examples [OMVCC and Caching](https://forum.treode.com/t/omvcc-and-http-caching/62).
+
+#### Read Requests  
+
+The cache message to the DB includes the table id, the desired key, and headers restricting the timestamps of the value.  (We describe in the Use of HTTP Headers section how each header restricts the value we read.)
+
+HTTP read requests should follow this format:
+
+    GET /table_id_string/key_string HTTP/1.1
+    Cache-Control: max-age=[max_age], no-cache
+    Read-TxClock: [read_timestamp_in_milliseconds]
+
+* **Quetion (4)**: Did I get the format for these messages (read request, write request, read response, write response) exactly correct?
+
+Here is an example of an HTTP read request:
+
+    GET /movie/star-wars HTTP/1.1
+    Cache-Control: max-age=300, no-cache
+    Read-TxClock: 1421236153024853
+
+In this example, the user is requesting the value of key "star-wars" in the table "movie".  The value should be written before (or on) time 1421236153024853 milliseconds but not more than 300 seconds before time 1421236153024853.  "no-cache" indicates that the cache MUST make a request to the DB server to retrieve this value, even if it has a value satisfying the requirements.
```

Comment:
> not more than 300 seconds

I don't understand this part. If the user reads as-of time t=1421236153024853, and the value was written at time t-1, then the DB will return that value at t-1. If the reads as-of time t, and the last the the value was written before t happens to have been in 1972, then the DB will return the value from 1972.